### ISSUE TITLE
Add property to evaluate if app is in dark mode

### ISSUE THREAD

**Issue by freakboy3742** [beeware/toga#2841]
### What is the problem or limitation you are having?

Most operating systems now have a concept of "dark mode"; if you're manually setting colors for labels or other text, you need to know if you're in dark mode to be able to determine if a label color should be high or low contrast.

### Describe the solution you'd like

`toga.App.dark_mode` should return True if the app is in dark mode, False otherwise. 

It might make sense to return `None` on platforms where this can't be determined (e.g., the platform doesn't have an explicit concept of "dark mode". This would evaluate as false-y, while still be checked as `is None`.

### Describe alternatives you've considered

Ideally, color selection is something that would be done with a full style sheet with media selectors; however, that requires a full CSS implementation (see [Colosseum](https://github.com/beeware/colosseum)); and regardless, colosseum would still require an app's style applicator to provide the current dark/light mode as an input to the style process.

### Additional context

See #2830 for the original request.

This will require an implementation on every backend, calling native APIs to determine dark mode status.

**Comment by rtpg** [comment]
Checking this one out 👓

**Comment by hyuri** [comment]
To be clear, this has only been implemented for macOS/Cocoa, right?

**Comment by freakboy3742** [comment]
Correct. The API endpoint *exists* on every platform; however, it returns "unknown" on every platform except macOS. Reopening the ticket with specific platform flags on that basis

**Comment by dietervansteenwegen** [comment]
According to what I found, the following allows setting the preference for the  current application:

```python
from gi.repository import Gtk
settings = Gtk.Settings.get_default()
settings.set_property("gtk-application-prefer-dark-theme", True)
```

However, the [Gtk documentation](https://docs.gtk.org/gtk3/class.Settings.html) does not seem to have a property to get the current system-wide setting.

**Comment by systematicguy** [comment]
I am looking at the Windows aspect of this.

**Comment by systematicguy** [comment]
@freakboy3742  are we open to a solution where we would use the `darkdetect` package?
I have run a quick test in my helloworld app and it just worked :tm:

https://pypi.org/project/darkdetect/

    import darkdetect
    
    are_we_dark: bool = darkdetect.isDark()

**Comment by HalfWhitt** [comment]
> [@freakboy3742](https://github.com/freakboy3742) are we open to a solution where we would use the `darkdetect` package?

That's a neat package, but:
- It was last updated three years ago.
- It's specifically designed to work with no depedencies (system-specific or otherwise), which is why it uses `subprocess` to make system command line calls to get its information. We already have access to native system APIs set up in each of our backends, so we shouldn't need a workaround like that.

That said, we may be able to learn some things by looking through it.

**Comment by dietervansteenwegen** [comment]
I've looked at it and it is indeed pretty simple:
* on Windows it uses `winreg` to read from the registry. 
* on Gtk it uses `subprocess` to query `gsettings`
* on Mac I'm out of my element and don't readily understand what's happening

It also includes a listener that accepts a callable to signal dark/light mode changes. 

This is something I could likely do (_as it is not much more than copy/paste_), though it would likely require a couple of reviews to get it as/where you'd want/need it to be. Also, leaving on holiday for a couple of weeks.

**Comment by HalfWhitt** [comment]
Well, we've already got Gtk and macOS, but we do still need it for Windows.

Toga's Windows backend includes [Python.NET](https://github.com/pythonnet/pythonnet) as a dependency; I haven't touched it myself, but I wouldn't be surprised if it includes a less verbose way to access those registry keys.

**Comment by ezzygarmyz** [comment]
> Well, we've already got Gtk and macOS, but we do still need it for Windows.
> 
> Toga's Windows backend includes [Python.NET](https://github.com/pythonnet/pythonnet) as a dependency; I haven't touched it myself, but I wouldn't be surprised if it includes a less verbose way to access those registry keys.

on Windows, the theme setting is stored in the registry, and it can be accessed quite simply.
Here's a minimal example using the standard library :


```python
import winreg as reg

def get_windows_theme():
    try:
        key_path = r"Software\Microsoft\Windows\CurrentVersion\Themes\Personalize"
        with reg.OpenKey(reg.HKEY_CURRENT_USER, key_path) as key:
            value, _ = reg.QueryValueEx(key, "AppsUseLightTheme")
            return "Light" if value == 1 else "Dark"
    except FileNotFoundError:
        return None
```

**Comment by ezzygarmyz** [comment]
> > Well, we've already got Gtk and macOS, but we do still need it for Windows.
> > Toga's Windows backend includes [Python.NET](https://github.com/pythonnet/pythonnet) as a dependency; I haven't touched it myself, but I wouldn't be surprised if it includes a less verbose way to access those registry keys.
> 
> on Windows, the theme setting is stored in the registry, and it can be accessed quite simply. Here's a minimal example using the standard library :
> 
> import winreg as reg
> 
> def get_windows_theme():
>     try:
>         key_path = r"Software\Microsoft\Windows\CurrentVersion\Themes\Personalize"
>         with reg.OpenKey(reg.HKEY_CURRENT_USER, key_path) as key:
>             value, _ = reg.QueryValueEx(key, "AppsUseLightTheme")
>             return "Light" if value == 1 else "Dark"
>     except FileNotFoundError:
>         return None

or using pythonnet to detect when theme is changed :
example :

```python
import clr
import time

clr.AddReference("System.Windows.Forms")

from Microsoft.Win32 import SystemEvents, Registry
from System import EventHandler

last_known_theme = None

def get_app_theme():
    """Reads the AppsUseLightTheme registry key (1=light, 0=dark)"""
    key = Registry.CurrentUser.OpenSubKey(
        r"Software\Microsoft\Windows\CurrentVersion\Themes\Personalize"
    )
    value = key.GetValue("AppsUseLightTheme", 1)
    return "Light" if value == 1 else "Dark"

def on_user_preference_changed(sender, e):
    global last_known_theme
    new_theme = get_app_theme()

    if new_theme != last_known_theme:
        last_known_theme = new_theme
        print(f"Theme changed! Current theme is: {new_theme}")

last_known_theme = get_app_theme()
print(f"Initial theme: {last_known_theme}")
print("Listening for theme changes... (Press Ctrl+C to exit)")

handler = EventHandler(on_user_preference_changed)
SystemEvents.UserPreferenceChanged += handler

try:
    while True:
        time.sleep(1)
except KeyboardInterrupt:
    print("Exiting...")
finally:
    SystemEvents.UserPreferenceChanged -= handler
```

**Comment by HalfWhitt** [comment]
I don't think we've formalized a core API yet for a listener on dark/light change, but it seems like it could be useful. @ezzygarmyz would you be interested in submitting a PR for the Windows detection, and/or a dark mode listener?

**Comment by ezzygarmyz** [comment]
> I don't think we've formalized a core API yet for a listener on dark/light change, but it seems like it could be useful. [@ezzygarmyz](https://github.com/ezzygarmyz) would you be interested in submitting a PR for the Windows detection, and/or a dark mode listener?

I really want to do that but I'm not a good in using github... I want to share my experience with you, maybe it will be useful for Toga users
Since the topic is about UI mode, I want to add another example for Windows on how to change the title bar from dark to light

```python

import ctypes
import clr
clr.AddReference("System.Windows.Forms")
import System.Windows.Forms as Forms

import toga
from toga.style.pack import Pack
from toga.colors import BLACK, WHITE
from toga.constants import CENTER

class UiMode(toga.App):
    def startup(self):

        dark_button = toga.Button(
            "Dark Mode",
            style=Pack(
                color = WHITE,
                background_color = BLACK,
                font_size=14,
                flex = 1,
                margin=10
            ),
            on_press=self.set_dark_mode
        )
        light_button = toga.Button(
            "Light Mode",
            style=Pack(
                font_size=14,
                flex = 1,
                margin=10
            ),
            on_press=self.set_light_mode
        )
        self.main_box = toga.Box(
            style=Pack(
                flex = 1,
                alignment = CENTER
            )
        )

        self.main_box.add(
            dark_button,
            light_button
        )

        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = self.main_box
        self.main_window.show()

    
    async def set_dark_mode(self, button):
        self.main_box.style.background_color = BLACK
        self.apply_title_bar_mode(1)

    async def set_light_mode(self, button):
        self.main_box.style.background_color = WHITE
        self.apply_title_bar_mode(0)


    def apply_title_bar_mode(self, ui_mode:int):
        try:
            hwnd = self.main_window._impl.native.Handle.ToInt32()
            value = ctypes.c_int(ui_mode)
            ctypes.windll.dwmapi.DwmSetWindowAttribute(
                ctypes.wintypes.HWND(hwnd),
                ctypes.c_int(20),
                ctypes.byref(value),
                ctypes.sizeof(value)
            )
            # refresh the current window to apply
            self.main_window._impl.native.WindowState = Forms.FormWindowState.Minimized
            self.main_window._impl.native.WindowState = Forms.FormWindowState.Normal
            self.main_window._impl.native.Activate()
        except Exception as e:
            print(f"Failed to apply ui mode: {e}")


def main():
    return UiMode()
```

**Comment by johnzhou721** [comment]
iOS:  ``UIView.traitCollection.userInterfaceStyle``

Also for some platforms the UI dark/light can autochange, so we need to probably implement something similar to a "color set" specifying different colors to display under dark/light mode differently.

**Comment by freakboy3742** [comment]
> Also for some platforms the UI dark/light can autochange, so we need to probably implement something similar to a "color set" specifying different colors to display under dark/light mode differently.

That falls into the domain of style directives, and having conditional style representations.

**Comment by johnzhou721** [comment]
I've created #3696 to track this since I realize that is sort of a separate feature.  I'm going to mark my comment as off-topic since there is now another place to track this.
